Content.Server/UserInterface/ActivatableUISystem.cs

```diff
@@ -0,0 +1,143 @@
+        }
+
+        private void OnActivate(EntityUid uid, ActivatableUIComponent component, ActivateInWorldEvent args)
+        {
+            if (args.Handled) return;
+            if (component.InHandsOnly) return;
+            args.Handled = InteractInstrument(args.User, component);
+        }
+
+        private void OnUseInHand(EntityUid uid, ActivatableUIComponent component, UseInHandEvent args)
+        {
+            if (args.Handled) return;
+            args.Handled = InteractInstrument(args.User, component);
+        }
+
+        private bool InteractInstrument(IEntity user, ActivatableUIComponent aui)
+        {
+            if (!user.TryGetComponent(out ActorComponent? actor)) return false;
+
+            if (aui.AdminOnly && !_adminManager.IsAdmin(actor.PlayerSession)) return false;
+
+            if (!_actionBlockerSystem.CanInteract(user.Uid))
+            {
+                user.PopupMessageCursor(Loc.GetString("base-computer-ui-component-cannot-interact"));
+                return true;
+            }
+
+            var ui = aui.UserInterface;
+            if (ui == null) return false;
+
+            if (aui.SingleUser && (aui.CurrentSingleUser != null) && (actor.PlayerSession != aui.CurrentSingleUser))
+            {
+                // If we get here, supposedly, the object is in use.
+                // Check with BUI that it's ACTUALLY in use just in case.
+                // Since this could brick the object if it goes wrong.
+                if (ui.SubscribedSessions.Count != 0) return false;
+            }
+
+            // If we've gotten this far, fire a cancellable event that indicates someone is about to activate this.
+            // This is so that stuff can require further conditions (like power).
+            var oae = new ActivatableUIOpenAttemptEvent(user);
+            RaiseLocalEvent(aui.OwnerUid, oae, false);
+            if (oae.Cancelled) return false;
+
+            SetCurrentSingleUser(aui.OwnerUid, actor.PlayerSession, aui);
+            ui.Toggle(actor.PlayerSession);
+            return true;
+        }
+
+        public void SetCurrentSingleUser(EntityUid uid, IPlayerSession? v, ActivatableUIComponent? aui = null)
+        {
+            if (!Resolve(uid, ref aui))
+                return;
+            if (!aui.SingleUser)
+                return;
+            if (aui.CurrentSingleUser != null)
+                aui.CurrentSingleUser.PlayerStatusChanged -= aui.OnPlayerStatusChanged;
+
+            aui.CurrentSingleUser = v;
+
+            if (v != null)
+                v.PlayerStatusChanged += aui.OnPlayerStatusChanged;
+            RaiseLocalEvent(uid, new ActivatableUIPlayerChangedEvent(), false);
+        }
+
+        public override void Update(float frameTime)
+        {
+            base.Update(frameTime);
+
+            foreach (var component in EntityManager.EntityQuery<ActivatableUIComponent>(true))
+            {
```

Comment:
If this system is going to eventually be used for all entity's with interact-to-open UI's, then for performance reasons this should probably not do an `EntityQuery` every tick. It should probably use something like the current `UserInterfaceSystem`, where it keeps an `activeInterfaces` list and just iterates over that. Then again, I'm really sure whether this performance impact would even be significant.

Reply:
The main problem is that the semantics of BoundUIClosedEvent make it difficult to determine when there is nobody observing the interface, `_isActive` is inaccessible and set too late...